File: seq2c2fm.pl
#!/usr/bin/env perl

use Getopt::Std;
use strict;

our ($opt_n, $opt_d, $opt_a, $opt_e, $opt_A, $opt_D, $opt_N, $opt_p, $opt_g, $opt_P, $opt_H, $opt_G, $opt_k, $opt_m, $opt_M);

getopts('HkgMn:d:a:e:D:A:N:p:P:G:m:') || USAGE();
$opt_H && USAGE();

my %purity;
if ( $opt_p ) {
    setPurity($opt_p);
}

my %mad;
if ( $opt_m ) {
    setMAD($opt_m);
}

my $MINDEL = $opt_D ? $opt_D : -2.0;
my $MINAMP = $opt_A ? $opt_A : 1.45;  # 6 copies for pure sample

my $MINEXONDEL = $opt_d ? $opt_d : -2.0;
my $MINEXONAMP = $opt_a ? $opt_a : 1.75;
my $MINEXON = $opt_e ? $opt_e : 1;
my $N = $opt_N ? $opt_N : 5; # If a breakpoint is called more than $N samples, then it's deemed a false positive and filtered
my @GAIN = $opt_G ? split(/:/, $opt_G) : ("MYC");
my %GAIN = map { ($_, 1); } @GAIN;
my $MAD = $opt_M ? $opt_M : 3.0;

my %count;
my %samples;
if ( $opt_k ) {
    print join("\t", qw(Sample Empty Variant_Type Gene NA - - Segment - - Transf_LogRatio Segments LogRatio alteration - - - - - - - Alteration));
    print "\n";
}
while( <> ) {
    my @a = split(/\t/);
    my $gene = $a[1];
    my $sample = $a[0];
    if ( $opt_n ) {
	if ( $a[0] =~ /$opt_n/ ) {
	    $sample = $1; 
	} else {
	    next;
	}
    }
    $samples{ $sample } = 1;
    my ($SMINDEL, $SMINAMP, $SMINEXONDEL, $SMINEXONAMP) = ($MINDEL, $MINAMP, $MINEXONDEL, $MINEXONAMP);
    my $pur = 1.0;
    if ( $purity{ $sample } ) {
	$pur = $purity{ $sample };
    } elsif ( $opt_P ) {
	$pur = $opt_P > 1 ? $opt_P/100.0 : $opt_P;
    }
    my $lr = $a[6];

    # If MAD is available, CNV will only be called if the values exceed the background
    if ( $opt_m ) {
	next unless( abs($lr) > $MAD*$mad{ $sample } );
    }

    if ( $opt_p || $opt_P ) {
        my $copy = (2.0**$lr-1+$pur)*2.0/$pur;  # For absolute copy
	if ( $copy <= 0 ) { # to capture the cases where $lr will be really small for homozygous deletions
	    $lr = -10;
	} else {
	    $lr = log($copy/2)/log(2);
	}
    }
    #if ( $a[10] && $a[10] < $MINEXON ) {
	#$lr = $a[12];
        #next unless( $lr > $MINAMP || $lr < $MINDEL );
    #}
    my $desc = ($a[8] && $a[8] eq "BP") ? "$a[10] of $a[11]" : "$a[11] of $a[11]";
    if ( $opt_g || $GAIN{ $gene } ) { # Only do whole gene for copy gains
        if ( $lr >= 0.75 && $lr < $SMINAMP ) {
	    print join("\t", $sample, "", "copy-number-alteration", $a[1], "NA", "-", "-", "$a[2]:$a[3]", "-", "-", sprintf("%.1f", 2**$lr*2), $desc, $lr, "gain", "-", "-", "-", "-", "-", "-", "-", "Gain"), "\n";
	    next;
	}
    }
    my $type = $lr < $SMINDEL ? "Deletion" : "Amplification";
    if ( $a[10] && $a[8] eq "BP" && ($a[10] >= $MINEXON || ($a[11] - $a[10]) >= $MINEXON) ) {
	$lr = $a[12];
	my $seg = $1 if ( $a[14] =~ /^(\S+)\(/ );
	$seg = "$1-$2" if ( $seg =~ /^(\d+).*,(\d+)$/ );
	if ( $a[9] eq "Del" ) {
	    $type = "Deletion";
	    $desc = "Del seg $seg";
	} elsif ( $a[9] eq "Dup" ) {
	    $type = "Duplication";
	    $desc = "Dup seg $seg";
	    $lr = $a[13];  # use the difference instead of absolute log2 ratio for duplications
	}
        next unless( $lr >= $SMINEXONAMP || $lr <= $SMINEXONDEL );
    } else {
	next unless( $lr >= $SMINAMP || $lr <= $SMINDEL );
    }
    next if ( $a[15] && $a[15] >= $N );
    if ( $type eq "Duplication" ) {
	print join("\t", $sample, "", "rearrangement", $a[1], "likely", "-", "-", "$a[2]:$a[3]", "-", "-", sprintf("%.1f", 2**$lr*2), $desc, $lr, "-", $a[1], $a[1], $desc, "-", "-", "-", "-", "Rearrangement"), "\n";
    } else {
	print join("\t", $sample, "", "copy-number-alteration", $a[1], "NA", "-", "-", "$a[2]:$a[3]", "-", "-", sprintf("%.1f", 2**$lr*2), $desc, $lr, $type eq "Deletion" ? "loss" : "amplification", "-", "-", "-", "-", "-", "-", "-", $type), "\n";
    }
}

# Set the tumor purity for each sample
sub setPurity {
    my $file = shift;
    open(PUR, $file);
    while( <PUR> ) {
        chomp;
	my ($sample, $purity) = split(/\t/);
	$purity /= 100.0 if ( $purity > 1 );
	$purity{ $sample } = $purity;
    }
    close(PUR);
}

# Set the tumor MAD for each sample
sub setMAD {
    my $file = shift;
    open(MAD, $file);
    while( <MAD> ) {
        chomp;
	my ($sample, $mad) = split(/\t/);
	$mad{ $sample } = $mad;
    }
    close(MAD);
}

sub USAGE {
    print STDERR<<USAGE;
    $0 [-k] [-g] [-e exons] [-n reg] [-N num] [-A amp] [-a amp] [-D del] [-d del] [-p purity_file] [-P purity] lr2gene_output

    The program will parse seq2c output and make calls for each gene and output in the format compatible with OncoPrint.
    It has the option to provide the purity so that log2ratio thresholds will be adjusted accordingly.  By default, it calls
    genes that are homozygously deleted or amplified with >= 6 copies.

    Options:
    -k Print header

    -g Whether to output copy gains [4-5] copies.  Default: no.  Use option -G to specify genes.  Should be used rarely.

    -p file
       A file contains the tumor purity for all samples.  Two columns, first is sample name, second is the purity in % or fraction [0-1].

    -P double
       The purity.  Default: 1 or 100%, as is for cell lines.  If set, all samples will assume to have the same purity.

    -m MAD_file
       A file contains the MAD values for all samples.  Two columns, first is the sample name, 2nd is the MAD in log2 fraction.

    -M MAD
       The MAD values.  Only raw log2ratio exceed MAD*(MAD value in -m file) will be considered.  The purpose of this option is to reduce false positives when the sample is vary noisy (high MAD values).  Default: 3, or similar to 3 standard deviation

    -n regex
       The regular expression to extract sample names.  Default: none.

    -N num
       If an breakpoint has been called in >= num of samples, it's deemed false positive.  Default: 5

    -e exons
       Minimum number of exons/amplicon.  Default: 1

    For whole gene:
    -D log2ratio
       The log2ratio to determine that a gene is homozygously deleted.  Default: -2.0

    -A log2ratio
       The log2ratio to determine that a gene is amplified.  Default: 1.45.

    For < 3 exons:
    -d log2ratio
       The minimum log2ratio to determine that 1-2 exons are deleted. Should be lower than [-d] to reduce false positives. Default: -2.5

    -a log2ratio
       The minimum log2ratio to determine that 1-2 exons are amplified. Should be larger than [-a] to reduce false positives. Default: 1.75
       
    For gains:
    -G Genes
       List of genes, seperated by ":", for which gain will be captured.  Default: MYC

AUTHOR
    Written by Zhongwu Lai

USAGE
    exit(0);
}

File: seq2c2fm.py
import argparse
from argparse import HelpFormatter
import re
from math import log

def set_purity(file: str) -> dict:
    """
    Set the tumor purity for each sample

    :param file: input file
    :type file: str
    :return: dict of sample purities
    :rtype: dict
    """
    purities = {}
    with open(file, 'r') as f:
        for line in f:
            sample, purity = line.strip().split('\t')
            purity = float(purity)
            if purity > 1:
                purity /= 100.0
            purities[sample] = purity
    return purities


def set_mad(file: str) -> dict:
    """
    Set the tumor MAD for each sample

    :param file: input file
    :return: dict of sample MADs
    """
    mads = {}
    with open(args.mad_file, 'r') as f:
        for line in f:
            sample, mad = line.strip().split('\t')
            mads[sample] = float(mad)
    return mads


def seq2c2fm(args) -> str:
    """
    Convert Seq2C log2gene output to OncoPrint output

    :param args: main program input argument
    :type args: argparse.Namespace
    :return: OncoPrint output
    :rtype: str
    """
    purity = {}
    if args.purity_file:
        purity = set_purity(args.purity_file)

    mad = {}
    if args.mad_file:
        mad = set_mad(args.mad_file)

    MINDEL = args.min_del
    MINAMP = args.min_amp  # 6 copies for pure sample
    MINEXONDEL = args.min_exon_del
    MINEXONAMP = args.min_exon_amp
    MINEXON = args.min_exon
    N = args.num  # If a breakpoint is called more than N samples, then it's deemed a false positive and filtered
    genes_gain = args.genes_gain.split(':')
    MAD = args.mad

    cols = ['Sample', 'Empty', 'Variant_Type', 'Gene', 'NA', '-', '-', 'Segment', '-', '-', 'Transf_LogRatio',
            'Segments', 'LogRatio', 'alteration', '-', '-', '-', '-', '-', '-', '-', 'Alteration']
    output = '\t'.join(cols) + '\n' if args.print_header else ''

    samples = {}
    with open(args.in_file, 'r') as f:
        for line in f:
            a = line.replace('\n', '').split('\t')
            sample, gene = a[0], a[1]
            if sample == 'Sample':  # skip header
                continue

            if args.reg:
                m = re.search(args.reg, sample)
                if m is not None:
                    sample = m.group(0)
                else:
                    continue
            samples[sample] = 1
            SMINDEL, SMINAMP, SMINEXONDEL, SMINEXONAMP = MINDEL, MINAMP, MINEXONDEL, MINEXONAMP

            pur = 1.0
            if sample in purity:
                pur = purity[sample]
            elif args.purity_percent:
                pur = args.purity_percent / 100.0 if args.purity_percent > 1 else args.purity_percent

            lr_str = a[6]
            lr = float(lr_str)

            # If MAD is available, CNV will only be called if the values exceed the background
            if sample in mad and abs(lr) <= MAD * mad[sample]:
                continue

            if args.purity_file or args.purity_percent:
                copy = (2.0 ** lr - 1 + pur) * 2.0 / pur;  # For tumor absolute cooy: lr = log2((N*p+2*(1-p))/2)
                if copy <= 0:  # to capture the cases where lr will be really small for homozygous deletions
                    lr = -10
                    lr_str = str(lr)
                else:
                    lr = log(copy/2)/log(2)

            desc = f"{a[10]} of {a[11]}" if a[8] == "BP" else f"{a[11]} of {a[11]}"
            if args.output_gain or gene in genes_gain:  # Only do whole gene for copy gains
                if lr >= 0.75 and lr < SMINAMP:
                    vals = [sample, "", "copy-number-alteration", a[1], "NA", "-", "-", f"{a[2]}:{a[3]}", "-", "-",
                            f"{2 ** lr * 2:.1f}", desc, lr_str, "gain", "-", "-", "-", "-", "-", "-", "-", "Gain"]
                    output += "\t".join(vals) + "\n"
                    continue

            _type = "Deletion" if lr < SMINDEL else "Amplification"

            if a[10] and a[8] == "BP" and (float(a[10]) >= MINEXON or (float(a[11]) - float(a[10])) >= MINEXON):
                lr_str = a[12]
                lr = float(lr_str)
                seg = ''
                m = re.match(r"(\S+)\(", a[14])
                if m is not None:
                    seg = m.group(1)
                    m2 = re.match(r"(\d+).*,(\d+)$", seg)
                    if m2 is not None:
                        seg = f"{m2.group(1)}-{m2.group(2)}"

                if a[9] == "Del":
                    _type = "Deletion"
                    desc = f"Del seg {seg}"
                elif a[9] == "Dup":
                    _type = "Duplication"
                    desc = f"Dup seg {seg}"
                    lr_str = a[13]
                    lr = float(lr_str)  # use the difference instead of absolute log2 ratio for duplications

                if not (lr >= SMINEXONAMP or lr <= SMINEXONDEL):
                    continue
            else:
                if not (lr >= SMINAMP or lr <= SMINDEL):
                    continue

            if a[15] and int(a[15]) >= N:
                continue

            if _type == "Duplication":
                vals = [sample, "", "rearrangement", a[1], "likely", "-", "-", f"{a[2]}:{a[3]}", "-", "-",
                        f"{2 ** lr * 2:.1f}", desc, lr_str, "-", a[1], a[1], desc, "-", "-", "-", "-", "Rearrangement"]
            else:
                vals = [sample, "", "copy-number-alteration", a[1], "NA", "-", "-", f"{a[2]}:{a[3]}", "-", "-",
                        f"{2 ** lr * 2:.1f}", desc, lr_str,
                        "loss" if _type == "Deletion" else "amplification", "-", "-", "-", "-", "-", "-", "-", _type]
            output += "\t".join(vals) + "\n"

    return output


def main():
    usage = '''seq2c2fm.py [-g] [-e exons] [-n reg] [-N num] [-A amp] [-a amp] [-D del] [-d del] [-p purity_file] [-P purity] -i lr2gene_output -o oncoPrint_output'''
    parser = argparse.ArgumentParser(prog='seq2c2fm', description='Convertor of Seq2C results to OncoPrint format',
                                     epilog=f'''This program will parse seq2c output and make calls for each gene and output in the format compatible with OncoPrint\nUsage: {usage}''',
                                     formatter_class=HelpFormatter)  # argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('-k', '--print_header', action='store_true', help='Print header')
    parser.add_argument('-g', '--output_gain', action='store_true', default=False, help='Whether to output copy gains [4-5] copies')
    parser.add_argument('-p', '--purity_file', type=str,
                        help='A file that contains the tumor purity for all samples. Two columns, first is sample name, second is the purity in %% or fraction [0-1]')
    parser.add_argument('-P', '--purity_percent', type=float,
                        help='The purity. Default: 1 or 100%%, as is for cell lines. If set, all samples will assume to have the same purity')
    parser.add_argument('-n', '--reg', default=None,
                        help='The regular expression to extract sample names. Default: none')
    parser.add_argument('-m', '--mad_file', type=str,
                        help='A file contains the MAD values for all samples.  Two columns, first is the sample name, 2nd is the MAD in log2 fraction')
    parser.add_argument('-M', '--mad', type=float, default=3.0,
                        help='The MAD. Default: 3, as is for cell lines. If set, all samples will assume to have the same MAD')
    parser.add_argument('-N', '--num', type=int, default=5,
                        help='If an breakpoint has been called in >= num of samples, it is deemed false positive. Default: 5')
    parser.add_argument('-e', '--min_exon', type=int, default=1, help='Minimum number of exons/amplicon. Default: 1')
    parser.add_argument('-D', '--min_del', type=float, default=-2.0,
                        help='For whole gene: The log2ratio to determine that a gene is homozygously deleted. Default: -2.0')
    parser.add_argument('-A', '--min_amp', type=float, default=1.45,
                        help='For whole gene: The log2ratio to determine that a gene is amplified. Default: 1.45.')
    parser.add_argument('-d', '--min_exon_del', type=float, default=-2.0,
                        help='The minimum log2ratio to determine that 1-2 exons are deleted. Should be lower than [-d] to reduce false positives. Default: -2.5')
    parser.add_argument('-a', '--min_exon_amp', type=float, default=1.75,
                        help='The minimum log2ratio to determine that 1-2 exons are amplified. Should be larger than [-a] to reduce false positives. Default: 1.75')
    parser.add_argument('-G', '--genes_gain', type=str, default='MYC',
                        help='List of genes, seperated by ":", for which gain will be captured. Default: MYC')
    parser.add_argument('-i', '--in_file', required=True, help='input file that is the lr2gene output')
    parser.add_argument('-o', '--out_file', required=True, help='Output file')

    args = parser.parse_args()
    # print(args)
    output = seq2c2fm(args)
    with open(args.out_file, 'w') as fout:
        fout.write(output)


if __name__ == '__main__':
    main()
